### ISSUE TITLE
[TypeError: Cannot read property 'apply' of undefined] when I use `save`

### ISSUE THREAD

**Issue by vitahlin** [Automattic/mongoose#5038]
**Do you want to request a *feature* or report a *bug*?**

I want request a bug. 

**What is the current behavior?**

When I use `save` function to save mongodb , it will throw an error:
```c 
[2017-03-04 20:07:45.427] [ERROR] console - [TypeError: Cannot read property 'apply' of undefined]
TypeError: Cannot read property 'apply' of undefined
    at /Users/vitah/Downloads/Code/VitahPomeloServer/game-server/node_modules/mongoose/lib/services/model/applyHooks.js:135:26
    at Promise._execute (/Users/vitah/Downloads/Code/VitahPomeloServer/game-server/node_modules/bluebird/js/release/debuggability.js:300:9)
    at Promise._resolveFromExecutor (/Users/vitah/Downloads/Code/VitahPomeloServer/game-server/node_modules/bluebird/js/release/promise.js:483:18)
    at new Promise (/Users/vitah/Downloads/Code/VitahPomeloServer/game-server/node_modules/bluebird/js/release/promise.js:79:10)
    at wrappedPointCut (/Users/vitah/Downloads/Code/VitahPomeloServer/game-server/node_modules/mongoose/lib/services/model/applyHooks.js:115:23)
    at /Users/vitah/Downloads/Code/VitahPomeloServer/game-server/node_modules/thunkify/index.js:43:12
    at /Users/vitah/Downloads/Code/VitahPomeloServer/game-server/node_modules/co/index.js:136:8
    at thunkToPromise (/Users/vitah/Downloads/Code/VitahPomeloServer/game-server/node_modules/co/index.js:135:10)
    at toPromise (/Users/vitah/Downloads/Code/VitahPomeloServer/game-server/node_modules/co/index.js:119:55)
    at next (/Users/vitah/Downloads/Code/VitahPomeloServer/game-server/node_modules/co/index.js:99:29)
    at onFulfilled (/Users/vitah/Downloads/Code/VitahPomeloServer/game-server/node_modules/co/index.js:69:7)
    at process._tickDomainCallback (internal/process/next_tick.js:129:7)
```

**If the current behavior is a bug, please provide the steps to reproduce.**

My schema `roleSchema.js`:
```c
var mongoose = require('mongoose');
var moment = require('moment');

///////////////////////////////////////////////////////////

var roleSchema = new mongoose.Schema({
	// 角色id
	uid: {
		type: Number,
		required: true,
		unique: true,
		index: true
	},

	// 玩家等级
	lv: {
		type: Number,
		default: 1
	}
});

roleSchema.pre('save', function(next) {
	console.log('用户' + this.uid + '保存，时间：' + moment());
	next();
});

// 生成json格式内容的函数
if (!roleSchema.options.toJSON) {
	roleSchema.options.toJSON = {};
}
/* jshint unused:false */
roleSchema.options.toJSON.transform = function(doc, ret) {
	delete ret._id;
	delete ret.__v;
};

// 生成Role模型
mongoose.model('Role', roleSchema);
```

My model `roleModel.js`:
```c
var roleSchema = require('../schemas/roleSchema');
var mongoose = require('mongoose');
var Role = mongoose.model('Role');
var modelUtil = require('../util/modelUtil.js');

module.exports.getByUid = function(uid, cb) {
	modelUtil.getByUid(Role, uid, function(err, role_model) {
		if (!!err) {
			console.error(err);
			return cb(err);
		}

		cb(null, role_model);
	});
};
```

My `modelUtil.js`:
```c 
module.exports.getByUid = function(model, uid, cb) {
	// 创建成功后
	function afterCreate(err, schema) {
		cb(err, schema);
	}

	// 查询成功后
	function afterFindOne(err, schema) {
		if (!!err) {
			return cb(err);
		}
		if (!!schema) {
			return cb(null, schema);
		} else {
			// 如果没有对应表，则创建
			model.create({
				uid: uid
			}, afterCreate);
		}
	}

	model.findOne({
		uid: uid
	}, afterFindOne);
};
```

then I use mongoose to do this:
```c
var co = require('co');
var thunkify = require('thunkify');
var roleModel = require('../../../models/roleModel.js');
var code = require('../../../consts/code.js');

/////////////////////////////////////////////////////////////////

module.exports = function(app) {
	return new Handler(app);
};

var Handler = function(app) {
	this.app = app;
};

Handler.prototype.add = function(msg, session, next) {
	var uid = session.uid;
	if (!uid) {
		return next(null, {
			code: code.PARAM_ERROR
		});
	}

	var onDo = function*() {
		var role_model = yield thunkify(roleModel.getByUid)(uid);

		role_model.lv =3;

		yield thunkify(role_model.save)();

		return next(null, {
			code: code.OK
		});
	};

	var onError = function(err) {
		console.error(err);
		return next(null, {
			code: code.FAIL
		});
	};

	co(onDo).catch(onError);
};
```

I use `co` and `thunkyfy`,and when I try like that ,It will report an error which is ` [ERROR] console - [TypeError: Cannot read property 'apply' of undefined]`.

**I have been writing like it for months and it works very well.**
**But yestorday , I git clone this code in another machine , then it reported this error.**
**I don't know where the problem is. Because the same code run in two machines,but one is work well, another is throw error.**
**They are all Centos7,nodejs version and mongoose version are the same.**

**Please mention your node.js, mongoose and MongoDB version.**
Node.js version:v5.12.0
Mongodb version:3.4.2
mongoose version:4.8.5

Thanks for your help.

**Comment by varunjayaraman** [comment]
I'm sorry, do you have a simpler repro script? Also don't forget to set the mongoose Promise constructor.

**Comment by vkarpov15** [comment]
The issue is `thunkify()`. `thunkify(roleModel.getByUid)` won't work because then the `getByUid` function won't know what `this` is. You'd get the same error if you do `const fn = roleModel.getByUid; fn();`. Please read up on [function context](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/this) . I also wrote a bunch about this in [my ebook about generators and co](http://es2015generators.com/).

TLDR: do this

```javascript
var role_model = yield cb => roleModel.getByUid(uid, cb);
yield cb => role_model.save(cb);
```

if you really want to use thunks. But there's no need since mongoose functions return a promise anyway, so its better to just do this:

```javascript
var role_model = yield roleModel.getByUid(uid);
yield role_model.save();
```

**Comment by vitahlin** [comment]
Thanks all of you.
